hypothesis-python/src/hypothesis/internal/lambda_sources.py

```diff
@@ -0,0 +1,426 @@
+from inspect import Parameter
+from typing import Any, Callable, Optional
+from weakref import WeakKeyDictionary
+
+from hypothesis.internal import reflection
+from hypothesis.internal.cache import LRUCache
+
+# we have several levels of caching for lambda descriptions.
+# * LAMBDA_DESCRIPTION_CACHE maps a lambda f to its description _lambda_description(f).
+#   Note that _lambda_description(f) may not be identical to f as it appears in the
+#   source code file.
+# * LAMBDA_DIGEST_DESCRIPTION_CACHE maps _function_key(f) to _lambda_description(f).
+#   _function_key implements something close to "ast equality":
+#   two syntactically identical (minus whitespace etc) lambdas appearing in
+#   different files have the same key. Cache hits here provide a fast path which
+#   avoids ast-parsing syntactic lambdas we've seen before. Two lambdas with the
+#   same _function_key will not have different _lambda_descriptions - if
+#   they do, that's a bug here.
+# * AST_LAMBDAS_CACHE maps source code lines to a list of the lambdas found in
+#   that source code. A cache hit here avoids reparsing the ast.
+LAMBDA_DESCRIPTION_CACHE: MutableMapping[Callable, str] = WeakKeyDictionary()
+LAMBDA_DIGEST_DESCRIPTION_CACHE: LRUCache[tuple[Any], str] = LRUCache(max_size=1000)
+AST_LAMBDAS_CACHE: LRUCache[tuple[str], list[ast.Lambda]] = LRUCache(max_size=100)
+
+
+def extract_all_lambdas(tree):
+    lambdas = []
+
+    class Visitor(ast.NodeVisitor):
+
+        def visit_Lambda(self, node):
+            lambdas.append(node)
+            self.visit(node.body)
+
+    Visitor().visit(tree)
+    return lambdas
+
+
+def extract_all_attributes(tree):
+    attributes = []
+
+    class Visitor(ast.NodeVisitor):
+        def visit_Attribute(self, node):
+            attributes.append(node)
+            self.visit(node.value)
+
+    Visitor().visit(tree)
+    return attributes
+
+
+def _function_key(f, *, bounded_size=False):
+    """Returns a digest that differentiates functions that have different sources.
+
+    Either a function or a code object may be passed. If code object, default
+    arg/kwarg values are not recoverable - this is the best we can do, and is
+    sufficient for the use case of comparing nested lambdas.
+    """
+    try:
+        code = f.__code__
+        defaults_repr = repr((f.__defaults__, f.__kwdefaults__))
+    except AttributeError:
+        code = f
+        defaults_repr = ()
+    consts_repr = repr(code.co_consts)
+    if bounded_size:
+        # Compress repr to avoid keeping arbitrarily large strings pinned as cache
+        # keys. We don't do this unconditionally because hashing takes time, and is
+        # not necessary if the key is used just for comparison (and is not stored).
+        if len(consts_repr) > 48:
+            consts_repr = hashlib.sha384(consts_repr.encode()).digest()
+        if len(defaults_repr) > 48:
+            defaults_repr = hashlib.sha384(consts_repr.encode()).digest()
```

Comment:
Should it be?

```python
defaults_repr = hashlib.sha384(defaults_repr.encode()).digest()
```

Otherwise, it fails with 

```
            if len(consts_repr) > 48:
                consts_repr = hashlib.sha384(consts_repr.encode()).digest()
            if len(defaults_repr) > 48:
>               defaults_repr = hashlib.sha384(consts_repr.encode()).digest()
                                               ^^^^^^^^^^^^^^^^^^
E               AttributeError: 'bytes' object has no attribute 'encode'. Did you mean: 'decode'?
```

in one of the Schemathesis tests